Add ADR for support of NetworkPolicies in Istio module (#1869)
kyma-project/istio: docs/contributor/adr/0015-networkpolicies-istio-kyma-system.md

# NetworkPolicies for istio-system and kyma-system

## Status
Proposed

## Context
Clusters that enforce a default-deny NetworkPolicy require explicit allow rules for Istio components to install and operate.
This ADR captures those requirements so they can be consistently applied and maintained.

To ensure that Istio components can function properly under default-deny policies,

> we need to define the necessary NetworkPolicies for both `istio-system` and `kyma-system` namespaces.

Reviewer: ```suggestion
To ensure that Istio components function properly under default-deny policies,
we must define the necessary NetworkPolicies for both `istio-system` and `kyma-system` namespaces.
```

2. Allow API server access (TCP 443) for:
   - `istio-controller-manager` in `kyma-system`
   - `istiod` (`istio: pilot`) in `istio-system`
   - `istio-cni-node` in `istio-system`
3. Allow control-plane communication between `istiod` and the `istio-ingressgateway`:
   - Allow egress from `istio-ingressgateway` to `istiod` on port 15012 (TCP, XDS protocol).
   - Allow ingress to `istiod` on port 15012 from sidecars and `istio-ingressgateway`.
4. Allow external ingress to `istio-ingressgateway` on TCP 8080/8443 for traffic entering the cluster.
5. Allow ingress to the `istiod` webhook endpoint on TCP 15017 from the API server for validating and mutating operations.
6. Allow `istiod` egress to common JWKS endpoint ports (TCP 80/443) for external JWT verification.
> [!NOTE]
> Allowing traffic to the 443 and 80 is not necessarily sufficient for all cases, as some JWKS endpoints might be running on non-standard ports.
> If users have specific requirements for accessing JWKS endpoints on non-standard ports, it might be required to either allow users to create custom NetworkPolicies in

> > the `istio-system` namespace or to provide a way to specify additional allowed ports for `istiod` egress in the Istio CR.

Reviewer: ```suggestion
  > [!NOTE]
  > Allowing traffic to the 443 and 80 is not necessarily sufficient for all cases, as some JWKS endpoints might be running on non-standard ports.
  > If users have specific requirements for accessing JWKS endpoints on non-standard ports, it might be required to either allow users to create custom NetworkPolicies in
  > the `istio-system` namespace or to provide a way to specify additional allowed ports for `istiod` egress in the Istio CR.
```

> 7. Allow user-enabled egress from `istio-ingressgateway` to backend services by permitting egress to specifically labeled pods in user namespaces.

Reviewer: ```suggestion
- Allow user-enabled egress traffic from `istio-ingressgateway` to backend services by permitting egress to specifically labeled Pods in user namespaces.
```

To ensure that the policies are enforced as soon as user enables the setting the Istio module components will be restarted (rollout restart)

> to terminate already established TCP connections.

Reviewer: ```suggestion
To ensure that the policies are enforced as soon as the user enables the setting, the Istio module's components must be restarted (rollout restart)
to terminate already established TCP connections.
```

  - Egress
  egress:
  - ports:
    - protocol: TCP
      port: 80
    - protocol: TCP
      port: 443
```

6. Allow user-enabled egress traffic from `istio-ingressgateway` to backend services by creating a NetworkPolicy
that allows egress to specifically labeled pods in user namespaces:

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: kyma-project.io--istio-ingressgateway-egress
  namespace: istio-system
spec:
  podSelector:
    matchLabels:
      istio: ingressgateway
  policyTypes:
  - Egress
  egress:
  - to:
    - podSelector:
        matchLabels:
          networking.kyma-project.io/from-ingressgateway: allowed
      namespaceSelector: {}
```

### Restart Istio Components to Enforce Policies
To ensure that the newly applied NetworkPolicies take effect immediately, all Istio components will be restarted (rollout restart)
when the user enables the NetworkPolicy support flag in the Istio CR.
This will terminate any existing TCP connections and allow the new policies to be enforced without delay.
This means that a restart should happen whenever the value of `enableModuleNetworkPolicies` is changed from `false` to `true` or vice versa,

> to ensure that the correct policies are applied based on the user's choice.

Reviewer: ```suggestion
To ensure that the newly applied NetworkPolicies take effect immediately, all Istio components must be restarted (rollout restart)
when the user enables the NetworkPolicy support flag in the Istio CR.
Enabling the flag terminates any existing TCP connections and allows the new policies to be enforced without delay.
This means that a restart should happen whenever the value of **enableModuleNetworkPolicies** is changed from `false` to `true` or vice versa,
to ensure that the correct policies are applied based on the user's choice.
```

```

3. Allow the following for `istio-ingressgateway`:
- Allow egress access to `istiod` on port 15012 (XDS).
- Allow egress access to DNS.
- Allow ingress access from external traffic on ports 80/443.

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: kyma-project.io--istio-ingressgateway
  namespace: istio-system
spec:
  podSelector:
    matchLabels:
      istio: ingressgateway
  policyTypes:
  - Egress
  - Ingress
  egress:
  - to:
    - podSelector:
        matchLabels:
          istio: pilot
    ports:
    - protocol: TCP
      port: 15012
  - ports:
    - protocol: UDP
      port: 53
    - protocol: TCP
      port: 53
  ingress:
  - ports:
    - protocol: TCP
      port: 8080
    - protocol: TCP
      port: 8443
```

4. Allow egress access to DNS and APIServer for `istio-cni-node`:
```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: kyma-project.io--istio-cni-node
  namespace: istio-system
spec:
  podSelector:
    matchLabels:
      k8s-app: istio-cni-node
  policyTypes:
  - Egress
  egress:
  - ports:
    - protocol: UDP
      port: 53
    - protocol: TCP
      port: 53
  - ports:
    - protocol: TCP
      port: 443

> ```

Reviewer: ```suggestion
  ```yaml
  apiVersion: networking.k8s.io/v1
  kind: NetworkPolicy
  metadata:
    name: kyma-project.io--istio-cni-node
    namespace: istio-system
  spec:
    podSelector:
      matchLabels:
        k8s-app: istio-cni-node
    policyTypes:
    - Egress
    egress:
    - ports:
      - protocol: UDP
        port: 53
      - protocol: TCP
        port: 53
    - ports:
      - protocol: TCP
        port: 443
  ```
```
